### ISSUE TITLE
Macro expansion recurses endlessly?

### ISSUE THREAD

**Issue by Veykril** [rust-analyzer/rust-analyzer#10256]
The following macro + invocation(taken from https://github.com/rust-lang/rust/issues/88577#issuecomment-914665202) seems to recurse endlessly(or just takes a very long time?) even emitting a warning in name res, `[WARN hir_def::nameres::collector] macro expansion is too deep`
```rs
macro_rules! many_args {
    ([$($t:tt)*]#$($h:tt)*) => {
        many_args!{[$($t)*$($t)*]$($h)*}
    };
    ([$($t:tt)*]) => {
        fn _f($($t: ()),*) {}
    }
}

many_args!{[_]########## ######}
```

**Comment by WaffleLapkin** [comment]
I think this is not an issue anymore, for me `RA_LOG="hir_def=info" rust-analyzer analysis-stats .` doesn't produce any warnings and seems to be working fine?

![2022-06-05_14-17](https://user-images.githubusercontent.com/38225716/172045881-51f9a131-e941-45cc-b6d8-ef42bb67bbdf.png)

**Comment by Veykril** [comment]
it is giving me an expansion error now, so I assume we broke something in macro expansion again
![image](https://user-images.githubusercontent.com/3757771/172046156-ecb99d82-7059-4581-89a0-7564ed129621.png)

**Comment by WaffleLapkin** [comment]
The only thing I get is the error from rustc about too many arguments:
![2022-06-05_14-41](https://user-images.githubusercontent.com/38225716/172046687-46bbd44b-12e6-44ee-9f14-21a3ef8f9352.png)

**Comment by Veykril** [comment]
Huh? That seems odd, are you certain you are on the latest version? (or do you have macro-error disabled in your settings perhaps)

**Comment by WaffleLapkin** [comment]
r-a version: `815b43429 2022-06-05 nightly`
These are all the r-a settings:
```text
"rust-analyzer.updates.channel": "nightly",
"rust-analyzer.cargo.allFeatures": true,
"rust-analyzer.checkOnSave.enable": true,
"rust-analyzer.imports.prefix": "crate",
"rust-analyzer.cargo.buildScripts.enable": true,
"rust-analyzer.rustc.source": "discover",
```
Though it seems like the first two are outdated...
